phidgets/phidget.go

```diff
@@ -7,18 +7,126 @@ package phidgets
 #include <phidget22.h>
 */
 import "C"
+import (
+	"errors"
+	"fmt"
+	"time"
+	"unsafe"
+)
 
 //Phidget - general phidget interface that all phidgets are derived from (for ease of type management)
+type phidget struct {
+	handle C.PhidgetHandle
+	class  string
```

Comment:
Any reason to break have `class` defined here vs. making the C call when necessary?

Here's my diff that runs:

```
diff --git a/phidgets/phidget.go b/phidgets/phidget.go
index 00e191e..11aa4a9 100644
--- a/phidgets/phidget.go
+++ b/phidgets/phidget.go
@@ -17,7 +17,6 @@ import (
 //Phidget - general phidget interface that all phidgets are derived from (for ease of type management)
 type phidget struct {
        handle C.PhidgetHandle
-       class  string
 }
 
 type Phidget interface {
@@ -46,7 +45,6 @@ func (p *phidget) OpenWaitForAttachment(timeout time.Duration) error {
        if cerr := C.Phidget_openWaitForAttachment(p.handle, C.uint(timeout.Milliseconds())); cerr != C.EPHIDGET_OK {
                return p.phidgetError(cerr)
        }
-       p.class, _ = p.GetChannelClassName()
        return nil
 }
 
@@ -56,8 +54,9 @@ func (p *phidget) phidgetError(cerr C.PhidgetReturnCode) error {
        }
        var errorString *C.char
        C.Phidget_getErrorDescription(cerr, &errorString)
-       if p.class != "" {
-               return errors.New(fmt.Sprintf("%s: %s", p.class, C.GoString(errorString)))
+       class, _ := p.GetChannelClassName()
+       if class != "" {
+               return errors.New(fmt.Sprintf("%s: %s", class, C.GoString(errorString)))
        }
        return errors.New(C.GoString(errorString))
 }
```

Reply:
There's a possible recursive issue there - p.GetChannelClassName called phidgetError. That was the reason to cache it. But we could just call the C.Phidget_getChannelClassName directly instead, that would be fine.

Reply:
How about like this?
```
diff --git a/phidgets/phidget.go b/phidgets/phidget.go
index 00e191e..a96cb37 100644
--- a/phidgets/phidget.go
+++ b/phidgets/phidget.go
@@ -46,7 +46,6 @@ func (p *phidget) OpenWaitForAttachment(timeout time.Duration) error {
        if cerr := C.Phidget_openWaitForAttachment(p.handle, C.uint(timeout.Milliseconds())); cerr != C.EPHIDGET_OK {
                return p.phidgetError(cerr)
        }
-       p.class, _ = p.GetChannelClassName()
        return nil
 }

@@ -55,9 +54,10 @@ func (p *phidget) phidgetError(cerr C.PhidgetReturnCode) error {
                return nil
        }
        var errorString *C.char
+       var className *C.char
        C.Phidget_getErrorDescription(cerr, &errorString)
-       if p.class != "" {
-               return errors.New(fmt.Sprintf("%s: %s", p.class, C.GoString(errorString)))
+       if C.Phidget_getChannelClassName(p.handle, &className) == C.EPHIDGET_OK {
+               return errors.New(fmt.Sprintf("%s: %s", C.GoString(className), C.GoString(errorString)))
        }
        return errors.New(C.GoString(errorString))
```

Reply:
Yeah, I guess the trade-off is making an extra C call on every phidgetError vs a small memory increase per phidget struct. The new method protects us in any weird situation where we get out of sync with the C lib. I'd prefer to keep state there.

Reply:
Yes, agreed. I would imagine that asking for the class of a phidget is actually just a constant table lookup anyway, so I doubt we're really saving anything by doing our own caching. Just the C.char* -> GoString, but that's also probably pretty tiny.

Reply:
I think you're right, I disassembled the latest branches and it looks like it's doing a `MOVQ` on x86_64 once we hit the C lib:

```
TEXT github.com/jrcichra/gophidgets/phidgets._Cfunc_Phidget_getChannelClassName(SB) _cgo_gotypes.go
  _cgo_gotypes.go:903	0x70cde0		64488b0c25f8ffffff	MOVQ FS:0xfffffff8, CX												
  _cgo_gotypes.go:903	0x70cde9		483b6110		CMPQ 0x10(CX), SP												
  _cgo_gotypes.go:903	0x70cded		0f8682000000		JBE 0x70ce75													
  _cgo_gotypes.go:903	0x70cdf3		4883ec28		SUBQ $0x28, SP													
  _cgo_gotypes.go:903	0x70cdf7		48896c2420		MOVQ BP, 0x20(SP)												
  _cgo_gotypes.go:903	0x70cdfc		488d6c2420		LEAQ 0x20(SP), BP												
  _cgo_gotypes.go:903	0x70ce01		c744244000000000	MOVL $0x0, 0x40(SP)												
  _cgo_gotypes.go:904	0x70ce09		488d442430		LEAQ 0x30(SP), AX												
  _cgo_gotypes.go:904	0x70ce0e		4889442418		MOVQ AX, 0x18(SP)												
  _cgo_gotypes.go:904	0x70ce13		488b0dee363600		MOVQ github.com/jrcichra/gophidgets/phidgets._cgo_7ad3368ddf45_Cfunc_Phidget_getChannelClassName(SB), CX	
  _cgo_gotypes.go:904	0x70ce1a		48890c24		MOVQ CX, 0(SP)													
  _cgo_gotypes.go:904	0x70ce1e		4889442408		MOVQ AX, 0x8(SP)												
  _cgo_gotypes.go:904	0x70ce23		e8d8b2cfff		CALL runtime.cgocall(SB)											
  _cgo_gotypes.go:905	0x70ce28		803d11d03d0000		CMPB $0x0, runtime.cgoAlwaysFalse(SB)										
  _cgo_gotypes.go:905	0x70ce2f		750a			JNE 0x70ce3b													
  _cgo_gotypes.go:909	0x70ce31		488b6c2420		MOVQ 0x20(SP), BP												
  _cgo_gotypes.go:909	0x70ce36		4883c428		ADDQ $0x28, SP													
  _cgo_gotypes.go:909	0x70ce3a		c3			RET														
  _cgo_gotypes.go:906	0x70ce3b		488b442430		MOVQ 0x30(SP), AX												
  _cgo_gotypes.go:906	0x70ce40		48890424		MOVQ AX, 0(SP)													
  _cgo_gotypes.go:906	0x70ce44		e8d72fd0ff		CALL runtime.convT64(SB)											
  _cgo_gotypes.go:906	0x70ce49		488d05109b0900		LEAQ 0x99b10(IP), AX												
  _cgo_gotypes.go:906	0x70ce50		48890424		MOVQ AX, 0(SP)													
  _cgo_gotypes.go:906	0x70ce54		e8c7aecfff		CALL runtime.cgoUse(SB)												
  _cgo_gotypes.go:907	0x70ce59		488d05405a0900		LEAQ 0x95a40(IP), AX												
  _cgo_gotypes.go:907	0x70ce60		48890424		MOVQ AX, 0(SP)													
  _cgo_gotypes.go:907	0x70ce64		488b442438		MOVQ 0x38(SP), AX												
  _cgo_gotypes.go:907	0x70ce69		4889442408		MOVQ AX, 0x8(SP)												
  _cgo_gotypes.go:907	0x70ce6e		e8adaecfff		CALL runtime.cgoUse(SB)												
  _cgo_gotypes.go:907	0x70ce73		ebbc			JMP 0x70ce31													
  _cgo_gotypes.go:903	0x70ce75		e8a672d6ff		CALL runtime.morestack_noctxt(SB)										
  _cgo_gotypes.go:903	0x70ce7a		e961ffffff		JMP github.com/jrcichra/gophidgets/phidgets._Cfunc_Phidget_getChannelClassName(SB)	
```

Reply:
Granted, I  could have opened up the phidget library (open source), but that's not as much fun :)